### PR TITLE
VSOCK interface

### PR DESCRIPTION
**Pull request by skobyda**
#  Vsock device support

Virtual socket support enables communication between the host and guest over a socket. Cockpit Machines now has support for setting up such a device.

Screenshot from 2023-05-16 12-43-49

The user can choose to configure a custom identifier, or let have it assigned automatically upon a VM's boot. The identifier is used by the host to uniquely identify vsock of a specific guest.

Screenshot from 2023-05-16 12-44-02

Please note that vsock still requires special vsock-aware software (e.g. socat) to communicate over the socket.

### CODE REVIEW

**Comment by skobyda** [comment]
The vsock facilitates communication between virtual machines and the host they are running on independent of virtual machine network configuration. Vsock Context Identifier is used to define the "address" of the vsock, e.g. when you have multiple running VMs with open vsocks, you are able identify each vsock from the host's point of view.

- You can have only one vsock attached to the VM.
- You can choose "Assign automatically". In this situation, libvirt will choose the lowest currently unused context identifier (i.e. no running VM currently has open vsock with such CID)  upon a start of VM.
[vsock_assign_automatically.webm](https://user-images.githubusercontent.com/42733240/222157831-974443d3-1373-49ec-91b8-e49498f611bc.webm)

- Lower context identifier is 3. CIDs 0-2 are reserved and user will not be able to select them.
- It's possible to assign the same CID to multiple VMs, as long as user doesn't plan to run them at the same time.
[vsock_assign_manually.webm](https://user-images.githubusercontent.com/42733240/222157814-86916340-c027-440a-9325-cedab38b042a.webm)

- You can hotplug and unplug vsock to/from running VM, but you cannot edit vsock on a running VM. If you edit it, changes will take effect after rebooting the VM.
[vsock_running.webm](https://user-images.githubusercontent.com/42733240/222157796-8cbc212b-6028-4582-9a63-37305213c59c.webm)

**Review by garrett — CHANGES_REQUESTED** [review]
Wow, we really need to do something about "general" being a place to shove things. I still don't understand the difference between "general" and "hypervisor details" either.

Can we work on fixing this (outside of this PR) as soon as possible? It's really beyond the point of being a scaling problem.

OK, now back to this PR: I have no idea what this is or what it is supposed to be used for. Please explain why and how someone would want and use this. ("To enable vsock" is not an answer. It would need to cover an actual task someone would want to do and/or solve a problem that someone has.)

Marking as "request changes", as we'll need to figure out what this is and (at least) change the messaging around it.

**Comment by skobyda** [comment]
> OK, now back to this PR: I have no idea what this is or what it is supposed to be used for. Please explain why and how someone would want and use this. ("To enable vsock" is not an answer. It would need to cover an actual task someone would want to do and/or solve a problem that someone has.)

Generally, it's about setting up communication between the VM and host without having to setup the whole networking stack. 
I made a list of very simplified use cases:

**SSH**
You can make connecting to VM thru ssh or copying files work without setting up networking. Imagine you have a VM in an isolated state where setting up networking for VM is not an option. Or just setting up a whole networking stack for a simple ssh is too complex and hard to maintain. Virtual sock is simpler. You can just setup a virtual socket:
_You want to copy some files, e.g. network traffic dump from VM to a host_
1. Configure virtual socket thru cockpit-machines
2. Setup socket-listen
on VM: `sudo socat SOCKET-LISTEN: .....`
on host: `sudo socat TCP4-LISTEN:2222 .....`
3. Now you can ssh into the VM: ssh root@localhost -p 2222, you can transfer files thru scp, etc. between host and VM all without complex networking.

**Communication between custom programs**
Normally, if you had 2 custom programs running on your system, and you want them to communicate, you could setup TCP/UDP/whatever socket communication between them. Now image however that one of these processes/programs is running in VM. you then:
1. Configure virtual socket thru cockpit-machines
2. open in VM and host
3. Your programs can now communicate with each other

**Another big reason for vsock is network file system.**
Again, you have VM where setting up networking is not easy, but you want to setup NFS between your host and VM. You can achieve it:
1. Configure virtual socket thru cockpit-machines
2. On host, create the nfs mount point on the host and invoke [vsock-nfds deamon](https://github.com/stefanha/linux/blob/vsock-nfsd/go.sh) on host
3. In VM, mount the host's NFS share with options `clienaddr=CID,proto=vsock`, where CID specifies the Context identifier you have configured thru cockpit UI

**Comment by skobyda** [comment]
Fixed the last debian-testing flake, it should be green now

**Comment by skobyda** [comment]
@KKoukiou mind givind it code-review?

**Comment by skobyda** [comment]
All tests green

File: src/components/vm/overview/vsock.jsx
Comment (martinpitt) [inline src/components/vm/overview/vsock.jsx:114]: Please use "Vsock" consistently.

File: src/components/vm/overview/vsock.jsx
Comment (martinpitt) [inline src/components/vm/overview/vsock.jsx:185]: vsock vs. Vsock -- let's please pick one spelling.

File: src/libvirtApi/domain.js
Comment (martinpitt) [inline src/libvirtApi/domain.js:571]: What is that `1`? For spawn this must all be strings, no?

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks! Unfortunately there's a pixel test conflict. Code-wise this is mostly good, but I have some concerns about "selling" (explaining) this feature to users. Some documentation pointers about how to *use* it would really be nice. Like, some example how to SSH into a VM with a vsock.

File: src/components/vm/overview/vmOverviewCard.jsx
Comment (skobyda) [reply src/components/vm/overview/vmOverviewCard.jsx:225]: @garrett was not happy with messaging I use in the modal, so maybe we can wait for his mockup and figure out something there.
Btw Qemu's vsock is just another layer above Linux vsocks, so we can use [their documentation](https://man7.org/linux/man-pages/man7/vsock.7.html): ```
A socket address is defined as a combination of a 32-bit Context
       Identifier (CID) and a 32-bit port number.  The CID identifies
       the source or destination, which is either a virtual machine or
       the host.  The port number differentiates between multiple
       services running on a single machine.
```
The CID is the one that is configurable from the outside.

File: src/components/vm/overview/vsock.jsx
Comment (skobyda) [reply src/components/vm/overview/vsock.jsx:57]: What do you propose? To not be able to assign a CID to offline VM if some other VM also has the same CID? That's not what qemu, libvirt does, and I think we should not introduce this kind of check either.

In fact, I think we should not do any checks against non-running VMs in general.
Non-running VM is basically just a configuration template. You may want to migrate them to another host. Or some other VM might have been migrated to your host, and then you might end up with several VMs with the same CID anyway.  Or you may want one VM as backup of another, or as part of VM cluster. E.g. have one VM running, and another VM acting a backup of first VM if it fails, and thus they have to be configured the same way, and have same CID (even if only one is running at once).

A while back, were solving the same issue with disks, where we implemented disk sharing between multiple VMs for clustering purposes, which however cannot run at the same time. But they have same configuration, have disks attached, so if one VM fails, another can act as backup and access the same disk. Now, I don't know how many (if any) people would use vsocks in their cluster setups though, so maybe it's just a corner case.

But in summary, I think we should not prevent user from assigning the same CIDs to multiple VMs. We should warn them, but not prevent them.

File: src/libvirtApi/domain.js
Comment (skobyda) [reply src/libvirtApi/domain.js:583]: VM can have only 1 vsock attached to it, so this says to remove the one and only vsock on the machine.

Comment (skobyda) [reply src/libvirtApi/domain.js:583]: I can change it to string as `"1"` so it's not confusing, if you want

Comment (skobyda) [reply src/libvirtApi/domain.js]: Fixed

File: src/components/vm/overview/vsock.jsx
Comment (skobyda) [reply src/components/vm/overview/vsock.jsx:114]: Fixed, now it uses "vsock" everywhere

File: src/components/vm/overview/vsock.jsx
Comment (skobyda) [reply src/components/vm/overview/vsock.jsx]: Fixed

File: src/libvirtApi/domain.js
Comment (martinpitt) [reply src/libvirtApi/domain.js]: Yes, that's what I meant -- maybe there's some magic somewhere to convert all spawn arguments to strings, but it's confusing and brittle. Thanks!

File: src/components/vm/overview/vsock.jsx
Comment (martinpitt) [reply src/components/vm/overview/vsock.jsx:57]: I wasn't suggesting to not allow it, I was just curious about the use case -- in particular, why users would want to specify them manually and therefore introduce potential conflicts. Backup VMs seem moderately plausible.

> In fact, I think we should not do any checks against non-running VMs in general. 

The PR already does that, no? It needs to scan used CIDs for conflicts. I don't object, just pointing it out.

File: src/components/vm/overview/vmOverviewCard.jsx
Comment (martinpitt) [reply src/components/vm/overview/vmOverviewCard.jsx]: That manpage is the C interface though, also not user-facing documentation. I. e. it also does not explain how to actually use this and talk to the VM without a network. How to I use e.g. `ssh` or `nc` to talk to such a VM?

File: src/libvirtApi/domain.js
Comment (martinpitt) [reply src/libvirtApi/domain.js]: Note you didn't change it to `"1"`, but  dropped the argument entirely. That may explain the "ERROR    XML error: only a single vsock device is supported" test failure?

File: src/components/vm/overview/vsock.jsx
Comment (skobyda) [reply src/components/vm/overview/vsock.jsx:57]: > The PR already does that, no? It needs to scan used CIDs for conflicts. I don't object, just pointing it out.

I meant checks where we would then prevent users from doing some kind of action. We should throw a warning, but never prevent some action because of a theoretical conflict with some shutoff VM.

File: src/components/vm/overview/vmOverviewCard.jsx
Comment (skobyda) [reply src/components/vm/overview/vmOverviewCard.jsx]: > That manpage is the C interface though, also not user-facing documentation. I. e. it also does not explain how to actually use this and talk to the VM without a network. How to I use e.g. `ssh` or `nc` to talk to such a VM?

For to estabhlish connection using socat, you use socat's VSOCK-CONNECT address type, which requires CID.

Let's take this example:
You have multiple VMs. But you want to be able to setup connection between "MY_FAVORITE_VM" and the host:
1.  You configure vsock for MY_FAVORITE_VM and specify CID_VALUE = `42`
2.  Setup socket-listen on host, using previously configured CID_VALUE: socat VSOCK-LISTEN:1234 VSOCK-CONNECT:`42`:1234
3. Now we can launch nc from VM: nc --vsock -l 1234

File: src/components/vm/overview/vsock.jsx
Comment (skobyda) [reply src/components/vm/overview/vsock.jsx:185]: Unresolving: 
This doesn't look right. Here every entry should start with uppercase. So for overview tab, it should be `Vsock`

File: src/components/vm/overview/vsock.jsx
Comment (skobyda) [reply src/components/vm/overview/vsock.jsx:114]: Commented my opinion here: https://github.com/cockpit-project/cockpit-machines/pull/969#discussion_r1164467065

**Comment by KKoukiou** [comment]
@skobyda @garrett asked that we don't add new elements in vm overview before having the page redesgined. We already blocked https://github.com/cockpit-project/cockpit-machines/pull/1001 on the redesign and I guess this applies to this as well.

**Comment by skobyda** [comment]
> @skobyda @garrett asked that we don't add new elements in vm overview before having the page redesgined. We already blocked #1001 on the redesign and I guess this applies to this as well.

Yeah, this would not get merged without his approval. I was just trying to get at least initial code review so I know this is implementet generally in a good way, so we don't have to spend a lot of time on code review once the design is finished

**Comment by garrett** [comment]
> Virtual sock is _simpler_. You can _just_ setup a virtual socket

Unless I'm mistaken: I _strongly_ disagree. vsock requires kernel configuration on both the host and guest, special setup for the VM, and then special software on the host to interact with the VM, and knowledge about all of that. Right? Whereas networking is something that works by default, and needs a custom configuration (which can be tricky, but we really need to spend some team cycles to make this work better, including firewall integration for VMs).

I was thinking about how it could be implemented from a UI viewpoint yesterday and a bit again today. This is correct, right?

1. Kernel configuration. 
   - This is something we can't really reliably do from Cockpit, right? And if this isn't enabled — in both the host _and_ guest — then vsock won't work. 
   - We'd need some UI for this. Probably that it's not supported with an (i) icon that has a popover that says that it needs to be supported in the kernel and a kernel configuration option needs to be added, including an example of how to do it.
2. Configuration for the VM.
   - Handled in this PR, but needs some polishing up.
3. Running something special on the host to communicate with the VM.
   - I'm not sure where this goes. It's kind of out of scope to directly handle this in Cockpit, but we need to message about it in the UI, else this entire PR is basically pointless (as not knowing what to do after enabling it means that it would serve no purpose to enable it).

Right now, as it stands, this PR does not address:

- What vsock is
- What it enables
- Why you'd use it
- How you'd use it

We need to communicate this Cockpit's UI. I'm working on it. First task was to try to make space for this feature (and others), second is for me to understand it better, third is to improve the UI based on everything above.

I'll post a mockup that addresses a few of these issues soon.

**Comment by skobyda** [comment]
@garrett Updates:
- Added the info popover to the overview, and also added the same text into the dialog
- Added similar info popover to watchdog for consistency
- God rid of the part talking about kernel configuration, and do not mention kernel modules, since all of these things are handled by QEMU automatically, and we do not need any preparation for it

Screenshot from 2023-04-18 11-46-12
 
Screenshot from 2023-04-18 11-46-07

Screenshot from 2023-04-18 11-46-16

**Comment by garrett** [comment]
> Got rid of the part talking about kernel configuration, and do not mention kernel modules, 

thumbs up emoji 

> since all of these things are handled by QEMU automatically, and we do not need any preparation for it

party popper emoji

---

OK, just going from the screenshots, I see the following:

- The filled in blue (i) icons should be outline (?) icons and be on the label, not the value. (The mockup is wrong; sorry! We discussed it in the call though.) The mockup was about it not being supported, but you (rightfully) changed the text to just an explanation.
- Context identifier should probably be remade to be a checkbox to a manual identifier and a number. (And if you click in the disabled number input, it'd automatically enable it.)

---

I need to get this running locally to provide more feedback. But I'm glad we hashed it out to make it so much simpler already!

I should probably also make some revised mockups, especially vsock-specific ones. But I think it's looking pretty good already!

**Comment by skobyda** [comment]
>     * Context identifier should probably be remade to be a checkbox to a manual identifier and a number. 

Something like this?
Screenshot from 2023-04-19 18-45-57

> (And if you click in the disabled number input, it'd automatically enable it.)

I have no idea how to do that. Once the field is disabled, it doesn't register any onClick events. Do we do something similar somewhere else in cockpit so I could get inspired?

**Comment by skobyda** [comment]
Rebased

**Comment by skobyda** [comment]
@garrett Btw here is how it looks now with new question icons:
Screenshot from 2023-05-02 13-02-24

> I think maybe checkbox should refer to an automatic identifier. Otherwise, it's not really intuitive that there is option for it to be generated automatically. At least when I open the dialog, and it looks like in the screenshot above, nothing tells me that context identifier will be generated automatically. It kinda looks like I have to just check the checkbox and fill in the value in order to continue, otherwise vsock will not work.

What do you say about my proposal for negating the checkbox?
Screenshot from 2023-05-02 13-03-57
Screenshot from 2023-05-02 13-03-54





> > ```
> > * Context identifier should probably be remade to be a checkbox to a manual identifier and a number. 
> > ```
> 
> Something like this?
> 
> Screenshot from 2023-04-19 19-03-56
> 
> I think maybe checkbox should refer to an automatic identifier. Otherwise, it's not really intuitive that there is option for it to be generated automatically. At least when I open the dialog, and it looks like in the screenshot above, nothing tells me that context identifier will be generated automatically. It kinda looks like I have to just check the checkbox and fill in the value in order to continue, otherwise vsock will not work.
> 
> It kinda gives me the feeling how you just have to check "I agree to terms and conditions" when you register somewhere, otherwise you cannot proceed. Maybe it's a weird comparison, but it gives me that kind of feeling smile
> 
> Maybe we could negate the checkbox, to something like this? Screenshot from 2023-04-19 18-45-57
> 
> > (And if you click in the disabled number input, it'd automatically enable it.)
> 
> I have no idea how to do that. Once the field is disabled, it doesn't register any onClick events. Do we do something similar somewhere else in cockpit so I could get inspired?

**Comment by garrett** [comment]
That's not what it looks like in PatternFly. Here are examples from PF5:

image

https://pf5.patternfly.org/components/forms/form

PF4 is basically the same (the styling hasn't changed):

image

https://www.patternfly.org/v4/components/form

In your screenshot, the color and size are incorrect.

---

> What do you say about my proposal for negating the checkbox?

That isn't what I was talking about. :neutral_face: 

Here's what I was talking about:

add vsock - default

add vsock - override

We already do this elsewhere in Cockpit, such as in the "create container" dialog in cockpit-podman:

image

**Comment by skobyda** [comment]
And how it looks when somebody want's to assign the same identifier which is already assigned to other VM:
Screenshot from 2023-05-04 14-25-00

**Comment by skobyda** [comment]
Ahh, I just noticed I made the checkbox work the other way around, will fix it immediatelly

File: src/components/vm/overview/vsock.jsx
Comment (martinpitt) [reply src/components/vm/overview/vsock.jsx:178]: The description list now uses ["Vsock"](https://github.com/cockpit-project/cockpit-machines/pull/969/commits/d814abbd82597355dd008a74ae2c07b4194e25b3#diff-70cd150703f2ad775366f80db32dbd98dbeffb82408df0f1cb255aacafe87777R229), so that should be resolved now?

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks! Mostly LGTM now. I'm a bit sceptical about the debian test special case, this might be better as a naughty to avoid forgetting about it. But no strong opinion.

File: src/components/vm/overview/vmOverviewCard.jsx
Comment (garrett) [reply src/components/vm/overview/vmOverviewCard.jsx:56]: Yeah, it's tricky. "special vsock-aware software" might be better to be a little more specific without being too specific here. And docs would be nice, if we had some to link to.

**Comment by skobyda** [comment]
Updated the design. Made the VMs using the identifier bold. Changed the message to mention "vsock-aware software" and changed label to "Custom identifier". @garrett WDYT?
Screenshot from 2023-05-11 02-12-51

File: src/components/vm/overview/vmOverviewCard.jsx
Comment (martinpitt) [reply src/components/vm/overview/vmOverviewCard.jsx]: I disagree. "vsock-aware" software is obvious, but not helpful. Really, I insist: This needs a documentation URL or a *concrete* example how to use it -- like showing @skobyda 's `socat` example above. That's at least enough to give admins a hint what to google for and experiment with.

File: src/components/vm/overview/vsock.jsx
Comment (martinpitt) [inline src/components/vm/overview/vsock.jsx:63]: Translatable sentences must not be split. This is a common problem -- you *cannot* mix translation and HTML markup, at least not in this simple way. This imposes a grammatical structure which isn't true for all languages. If you feel strongly about the `<strong>`, then perhaps `_("Identifier in use by:")` and then the strong identifier  without the period (I think that would work with RTL). Otherwise, use cockpit.format() with a `$0`

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks!

File: src/components/vm/overview/vsock.jsx
Comment (garrett) [inline src/components/vm/overview/vsock.jsx:64]: I think this is missing a period to separate spaces.

```suggestion
            __html: remarkable.render(cockpit.format(_("Identifier in use by **$0**. VMs with an identical identifier cannot run at the same time."), vmsUsingCIDAddressString))
```

File: src/components/vm/overview/watchdog.jsx
Comment (garrett) [reply src/components/vm/overview/watchdog.jsx:38]: There's a merge conflict.

File: src/components/vm/overview/vsock.jsx
Comment (martinpitt) [reply src/components/vm/overview/vsock.jsx:64]: So this addresses my [earlier review comment](https://github.com/cockpit-project/cockpit-machines/pull/969#pullrequestreview-1421755844), but this also seems brittle to me: If translations manage to do anything wrong to these `**`, put them in unbalanced or an unexpected order, or translators try to do something else with them, or happen to use any more "magic" markdown character, this is going to be a mess and crash the page in the worst case.

If you want to do this, please check `git grep fmt_to_fragments` in cockpit -- that is in pkg/lib, you can use it in c-machines. This keeps the markup outside of the translatable bits, and is both much more robust as well as also more efficient than rendering through markdown.

Comment (skobyda) [reply src/components/vm/overview/vsock.jsx]: Fixed and used fmt_to_fragments

**Review by garrett — CHANGES_REQUESTED** [review]
It still has the error that if you click one button, it shows a spinner on both:

image

Additionally: What happens when you click cancel when the process is in action? (Shouldn't Cancel also be disabled? Same for the × icon.)

**Comment by skobyda** [comment]
> It still has the error that if you click one button, it shows a spinner on both:
> 
> image
> 
> Additionally: What happens when you click cancel when the process is in action? (Shouldn't Cancel also be disabled? Same for the × icon.)

Ah yeah, nice catch. Fixed that:
[Screencast from 2023-05-11 15-58-21.webm](https://github.com/cockpit-project/cockpit-machines/assets/42733240/748b225b-71d8-4741-8d2a-287f0b303ab3)

**Review by garrett — APPROVED** [review]
Thanks for solving the spinner problem.

Ideally, it'd check the checkbox when you click on the disabled number input and make cancel disabled whent he buttons are disabled too. But this is good enough and I'm going on PTO and I don't want to hold this up any longer. If you want to address those in follow-ups, that would be nice.

:+1: Thanks! Looks good!

File: src/components/vm/overview/vmOverviewCard.jsx
Comment (skobyda) [reply src/components/vm/overview/vmOverviewCard.jsx]: I could not find any red hat documentation. Then there is the [QEMU documentation](https://wiki.qemu.org/Features/VirtioVsock) or [Linux manual page](https://man7.org/linux/man-pages/man7/vsock.7.html) . But none of them are any more useful, as they only describe what vsock is (and how to create it), but not how to use it. I could not find any documentation which shows some example how to use it. I create [socat example](https://github.com/cockpit-project/cockpit-machines/pull/969#discussion_r1164462332) out of  some video demo from Devconf.

Comment (skobyda) [reply src/components/vm/overview/vmOverviewCard.jsx]: @martinpitt did you mean something like this:
Screenshot from 2023-05-16 11-55-22
Screenshot from 2023-05-16 11-55-09

Comment (martinpitt) [reply src/components/vm/overview/vmOverviewCard.jsx]: That's what I meant, yes. Really nice, thank you!

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks for the updates! This is that →← close from perfection. But let's let the current test run finish, to check for green-ness.

File: src/components/vm/overview/vsock.jsx
Comment (skobyda) [reply src/components/vm/overview/vsock.jsx:190]: Can I use the one with without colon, and then just add colon with `string_variable + ':'`, or that wouldn't work for some languages?

Comment (skobyda) [reply src/components/vm/overview/vsock.jsx]: In the end, decided to not use colon

Comment (martinpitt) [reply src/components/vm/overview/vsock.jsx]: No, colons may look different in other languages. Simplest case is RTL, where it wouldn't get appended.

**Comment by martinpitt** [comment]
Cheers! Code changes LGTM, nice to split it out into a helper and re-use it. Tests are still failing, once they get green, I'm happy!

**Review by martinpitt — APPROVED** [review]
Dakujem!

**Pull request merged**
